src/main/java/com/aws/greengrass/secretmanager/LocalStoreMap.java

```diff
@@ -0,0 +1,303 @@
+import com.aws.greengrass.deployment.DeviceConfiguration;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.secretmanager.crypto.Crypter;
+import com.aws.greengrass.secretmanager.crypto.KeyChain;
+import com.aws.greengrass.secretmanager.crypto.MasterKey;
+import com.aws.greengrass.secretmanager.crypto.RSAMasterKey;
+import com.aws.greengrass.secretmanager.exception.SecretCryptoException;
+import com.aws.greengrass.secretmanager.exception.SecretManagerException;
+import com.aws.greengrass.secretmanager.model.AWSSecretResponse;
+import com.aws.greengrass.secretmanager.model.SecretConfiguration;
+import com.aws.greengrass.secretmanager.model.SecretDocument;
+import com.aws.greengrass.secretmanager.store.FileSecretStore;
+import com.aws.greengrass.secretmanager.store.SecretStore;
+import com.aws.greengrass.security.SecurityService;
+import com.aws.greengrass.security.exceptions.ServiceUnavailableException;
+import com.aws.greengrass.util.Coerce;
+import com.aws.greengrass.util.LockFactory;
+import com.aws.greengrass.util.LockScope;
+import com.aws.greengrass.util.RetryUtils;
+import com.aws.greengrass.util.Utils;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyPair;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import javax.inject.Inject;
+
+import static com.aws.greengrass.deployment.DeviceConfiguration.DEVICE_PARAM_CERTIFICATE_FILE_PATH;
+import static com.aws.greengrass.deployment.DeviceConfiguration.DEVICE_PARAM_PRIVATE_KEY_PATH;
+
+public class LocalStoreMap {
+    private final Logger logger = LogManager.getLogger(LocalStoreMap.class);
+    private final ConcurrentHashMap<String, Labels> secrets;
+    private final SecretStore<SecretDocument, AWSSecretResponse> secretStore;
+    private final AtomicReference<Crypter> crypter = new AtomicReference<>();
+    private final SecurityService securityService;
+
+    private final Lock lock = LockFactory.newReentrantLock(this);
+
+    @Inject
+    LocalStoreMap(SecurityService securityService, FileSecretStore dao, DeviceConfiguration deviceConfiguration) {
+        this.secretStore = dao;
+        this.securityService = securityService;
+        deviceConfiguration.onAnyChange(((whatHappened, node) -> {
+            if (validUpdate(node, DEVICE_PARAM_CERTIFICATE_FILE_PATH) || validUpdate(node,
+                    DEVICE_PARAM_PRIVATE_KEY_PATH)) {
+                // TODO: If the device creds change, then cache and local store should be updated as well as the keys
```

Comment:
This todo is implemented, right?

It looks like if the device creds change, then we invalidate / reset the crypter so that the next time it's used we well create an up-to-date crypter at that point

If so, can we remove the TODO tag, but leave a version of the comment? (Because it is still useful; although I would mention how the local store will be updated, since it doesn't happen here)

Reply:
Updated the comment that we should try downloading latest secrets from cloud and encrypt them with new key. This will update the local store and cache proactively. Otherwise, the current code will do that per request only.